File: dsa-practice-ram/dsa-practice-ram/Arrays/SecondLargestThirdLargest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace dsa_practice_ram.Arrays
{
    internal class SecondLargestThirdLargest
    {
        public void FindSecondLargest()
        {
            int[] array = { 10, 20, 4, 45, 99, 99 };

            int firstLargest = int.MinValue;
            int secondLargest = int.MinValue;

            foreach (var num in array)
            {
                if (num > firstLargest)
                {
                    secondLargest = firstLargest;
                    firstLargest = num;
                }
                else if (num > secondLargest && num != firstLargest)
                {
                    secondLargest = num;
                }
            }

            Console.WriteLine(secondLargest);
        }

        public void FindThirdLargest()
        {
            int[] array = { 10, 20, 4, 45, 99, 99 };

            int firstLargest = int.MinValue;
            int secondLargest = int.MinValue;
            int thirdLargest = int.MinValue;

            foreach (var num in array)
            {
                if (num > firstLargest)
                {
                    thirdLargest = secondLargest;
                    secondLargest = firstLargest;
                    firstLargest = num;
                }
                else if (num > secondLargest && num < firstLargest) // handles if num == firstLargest
                {
                    thirdLargest = secondLargest;
                    secondLargest = num;
                }
                else if (num > thirdLargest && num < secondLargest)
                {
                    thirdLargest = num;
                }
            }

            if (thirdLargest == int.MinValue)
                Console.WriteLine("Third largest element not found (not enough distinct elements).");
            else
                Console.WriteLine(thirdLargest);
        }

        public void SecondSmallestElement()
        {
            int[] array = { 10, 2, 4, 20, 4, 45, 99, 99 };

            int firstSmallest = int.MaxValue;
            int secondSmallest = int.MaxValue;

            foreach (var num in array)
            {
                if (num < firstSmallest)
                {
                    secondSmallest = firstSmallest;
                    firstSmallest = num;
                }
                else if (num < secondSmallest && num != firstSmallest)
                {
                    secondSmallest = num;
                }
            }
            if (secondSmallest == int.MaxValue)
                Console.WriteLine("Second smallest element not found (not enough distinct elements).");
            else
                Console.WriteLine(secondSmallest);
        }

        public void FindThirdSmallest()
        {
            int[] array = { 10, 20, 4, 45, 2, 99, 99 };

            int firstSmallest = int.MaxValue;
            int secondSmallest = int.MaxValue;
            int thirdSmallest = int.MaxValue;

            foreach (var num in array)
            {
                if (num < firstSmallest)
                {
                    thirdSmallest = secondSmallest;
                    secondSmallest = firstSmallest;
                    firstSmallest = num;
                }
                else if (num < secondSmallest && num > firstSmallest) // handles if num == firstSmallest
                {
                    thirdSmallest = secondSmallest;
                    secondSmallest = num;
                }
                else if (num < thirdSmallest && num > secondSmallest)
                {
                    thirdSmallest = num;
                }
            }

            if (thirdSmallest == int.MaxValue)
                Console.WriteLine("Third smallest element not found (not enough distinct elements).");
            else
                Console.WriteLine(thirdSmallest);
        }
    }
}

File: MyLearning/PythonPractice/dsa_practice/arrays/second_largest_third_largest.py
"""Find second/third largest and smallest distinct elements."""

INT_MIN = -2 ** 31
INT_MAX = 2 ** 31 - 1

# C# SecondLargestThirdLargest.FindSecondLargest

def find_second_largest(array=None):
    """Return the second largest distinct element or INT_MIN if missing.

    Problem: Find the second largest distinct value in an array.
    Approach: Track first and second largest with a single pass.
    Time complexity: O(n)
    Space complexity: O(1)
    """
    if array is None:
        array = [10, 20, 4, 45, 99, 99]

    first_largest = INT_MIN
    second_largest = INT_MIN

    for num in array:
        if num > first_largest:
            second_largest = first_largest
            first_largest = num
        elif num > second_largest and num != first_largest:
            second_largest = num

    return second_largest


# C# SecondLargestThirdLargest.FindThirdLargest

def find_third_largest(array=None):
    """Return the third largest distinct element or None if missing.

    Problem: Find the third largest distinct value in an array.
    Approach: Track top three distinct values in one pass.
    Time complexity: O(n)
    Space complexity: O(1)
    """
    if array is None:
        array = [10, 20, 4, 45, 99, 99]

    first_largest = INT_MIN
    second_largest = INT_MIN
    third_largest = INT_MIN

    for num in array:
        if num > first_largest:
            third_largest = second_largest
            second_largest = first_largest
            first_largest = num
        elif num > second_largest and num < first_largest:
            third_largest = second_largest
            second_largest = num
        elif num > third_largest and num < second_largest:
            third_largest = num

    if third_largest == INT_MIN:
        return None
    return third_largest


# C# SecondLargestThirdLargest.SecondSmallestElement

def second_smallest_element(array=None):
    """Return the second smallest distinct element or None if missing.

    Problem: Find the second smallest distinct value in an array.
    Approach: Track first and second smallest in one pass.
    Time complexity: O(n)
    Space complexity: O(1)
    """
    if array is None:
        array = [10, 2, 4, 20, 4, 45, 99, 99]

    first_smallest = INT_MAX
    second_smallest = INT_MAX

    for num in array:
        if num < first_smallest:
            second_smallest = first_smallest
            first_smallest = num
        elif num < second_smallest and num != first_smallest:
            second_smallest = num

    if second_smallest == INT_MAX:
        return None
    return second_smallest


# C# SecondLargestThirdLargest.FindThirdSmallest

def find_third_smallest(array=None):
    """Return the third smallest distinct element or None if missing.

    Problem: Find the third smallest distinct value in an array.
    Approach: Track three smallest distinct values in one pass.
    Time complexity: O(n)
    Space complexity: O(1)
    """
    if array is None:
        array = [10, 20, 4, 45, 2, 99, 99]

    first_smallest = INT_MAX
    second_smallest = INT_MAX
    third_smallest = INT_MAX

    for num in array:
        if num < first_smallest:
            third_smallest = second_smallest
            second_smallest = first_smallest
            first_smallest = num
        elif num < second_smallest and num > first_smallest:
            third_smallest = second_smallest
            second_smallest = num
        elif num < third_smallest and num > second_smallest:
            third_smallest = num

    if third_smallest == INT_MAX:
        return None
    return third_smallest
